### PR TITLE
Use simulated dual-stack binds when using WSL

### LINKED ISSUE #12292: Cannot connect to debug port on Windows (WSL2)
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

I have been trying to migrate all Docker recipes to Podman for both Mac and Windows.
On Windows running Debian (WSL 2) I am able to install, build and run Podman container successfully, running Tomcat 7 service and access the service from Windows host (via localhost:8080).
However, I am not able to connect to debug port 8000 inside container for some reasons. (I'm using IntelliJ to debug)
**Steps to reproduce the issue:**

1. Run a CentOS 6 base instance with port mapping 8080 and 8000
```
podman run -td --name centos -p 8080:8080 -p 8000:8000 centos:6.10 bash -l
```

2. Install Java 1.8 and Tomcat 7

3. Setup Catalina option in /usr/share/tomcat/conf/tomcat.conf with the below option
```
CATALINA_OPTS="-Dlog.host=`hostname -s` -Xms256m -Xmx256m -XX:PermSize=256m -XX:MaxPermSize=256m -XX:+UseG1GC -Djava.security.egd=file:/dev/./urandom -Xdebug -Xrunjdwp:transport=dt_socket,address=8000,server=y,suspend=n"
```
4. Start tomcat service using `service tomcat start`
5. On Windows host, run IntelliJ, setup remote JVM debug to `Attach to remote JVM` to localhost:8000
6. Run the above debug configuration

**Describe the results you received:**
Error running 'remote-debug': Unable to open debugger port (localhost:8000): java.net.ConnectException "Connection refused: connect"

**Describe the results you expected:**
Debugger connects to Tomcat instance inside Podman container running on WSL 2.

**Additional information you deem important (e.g. issue happens only occasionally):**
Some additional things I want to add:
- I use the same recipe for Podman on Mac running Podman machine VM, and IntelliJ is able to connect to debug port localhost:8000
- I also tested with Windows Docker Desktop running WSL 2 backend and IntelliJ is also able to connect to debug port localhost:8000
- I checked if port 8000 is accessible from Windows host using netstat, and both port 8080 (obviously) and port 8000 were listening.
```
PS C:\Users\christopher.vo> netstat -a | findstr :8000                                                                    TCP    [::1]:8000             P50442:0               LISTENING
PS C:\Users\christopher.vo> netstat -a | findstr :8080                                                                    TCP    [::1]:8080             P50442:0               LISTENING
```

**Output of `podman version`:**
```
Version:      3.0.1
API Version:  3.0.0
Go Version:   go1.15.9
Built:        Thu Jan  1 09:00:00 1970
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.19.6
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: 'conmon: /usr/bin/conmon'
    path: /usr/bin/conmon
    version: 'conmon version 2.0.25, commit: unknown'
  cpus: 4
  distribution:
    distribution: debian
    version: "11"
  eventLogger: journald
  hostname: P50442
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.10.16.3-microsoft-standard-WSL2
  linkmode: dynamic
  memFree: 4746706944
  memTotal: 13265854464
  ociRuntime:
    name: crun
    package: 'crun: /usr/bin/crun'
    path: /usr/bin/crun
    version: |-
      crun version 0.17
      commit: 0e9229ae34caaebcb86f1fde18de3acaf18c6d9a
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    path: /tmp/podman-run-1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    selinuxEnabled: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: 'slirp4netns: /usr/bin/slirp4netns'
    version: |-
      slirp4netns version 1.0.1
      commit: 6a7b16babc95b6a3056b33fb45b74a6f62262dd4
      libslirp: 4.4.0
  swapFree: 4294967296
  swapTotal: 4294967296
  uptime: 5h 11m 13.78s (Approximately 0.21 days)
registries: {}
store:
  configFile: /home/christopher/.config/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 1
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: 'fuse-overlayfs: /usr/bin/fuse-overlayfs'
      Version: |-
        fusermount3 version: 3.10.3
        fuse-overlayfs: version 1.4
        FUSE library version 3.10.3
        using FUSE kernel interface version 7.31
  graphRoot: /home/christopher/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 27
  runRoot: /tmp/podman-run-1000/containers
  volumePath: /home/christopher/.local/share/containers/storage/volumes
version:
  APIVersion: 3.0.0
  Built: 0
  BuiltTime: Thu Jan  1 09:00:00 1970
  GitCommit: ""
  GoVersion: go1.15.9
  OsArch: linux/amd64
  Version: 3.0.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
Listing... Done
podman/stable,now 3.0.1+dfsg1-3+b2 amd64 [installed]
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Debian on WSL 2 (Windows)

### PR DESCRIPTION
**Pull request by n1hility**
Resolves a WSL problem where traffic from only one stack is relayed

Fixes #12292

Signed-off-by: Jason T. Greene <jason.greene@redhat.com>

### CODE REVIEW

**Comment by n1hility** [comment]
PTAL @Luap99

**Comment by n1hility** [comment]
@Luap99 Ok I switched it to modify the port mapping level instead. The reason I originally did this at the rootlet level was when I checked slrirp4netns netns was using rootlessport forwarding as well, so my thinking was 1 and 2 were covered and the issue wasn't problematic for the 3-4 rootful scenarios, so I was just keeping the change narrow. I take your point though that the networking implementations can evolve and expand so really should be handled earlier.

File: libpod/networking_linux.go [hunk 1]
```diff
@@ -63,18 +63,65 @@ const (
...
+	if len(host) == 0 {
+		host = "0.0.0.0"
+	}
+	ip := net.ParseIP(host)
+	splitLoopback := ip.IsLoopback() && ip.To4() == nil
+	// Map ::1 and 0.0.0.0/:: to ipv4 + ipv6 to simulate dual-stack
+	if ip.IsUnspecified() || splitLoopback {
+		ports = append(ports, port)
+		ports[0].Protocol = protocol + "4"
+		ports[1].Protocol = protocol + "6"
```
Comment (Luap99) [inline libpod/networking_linux.go:111]: This is not correct, types.PortMapping defines Protocol as comma separated list of protocols. This will never be the case in the libpod db because we deduplicate this when we create a container. However this is a implementation detail only we should not rely on that.

Also more important using tcp4/6 is not a supported protocol. I am sure that both cni and netavark will fail with that. Should'nt it be enough to bind both `::`  and `0.0.0.0` why do we need the extra 4/6. I think this is some golang implementation detail we should not use.

File: libpod/networking_linux.go [hunk 2]
```diff
@@ -63,18 +63,65 @@ const (
...
+	port.HostIP = ""
+	return []types.PortMapping{port}
+}
+
+func splitDualStack(port types.PortMapping) []types.PortMapping {
+	ports := []types.PortMapping{port}
+	protocol := port.Protocol
+	if strings.HasSuffix(protocol, "4") || strings.HasSuffix(protocol, "6") {
+		return ports
+	}
```
Comment (Luap99) [inline libpod/networking_linux.go:99]: This will never be the case see the other comment

File: libpod/networking_linux.go [hunk 3]
```diff
@@ -63,18 +63,65 @@ const (
...
+	ip := net.ParseIP(host)
+	splitLoopback := ip.IsLoopback() && ip.To4() == nil
+	// Map ::1 and 0.0.0.0/:: to ipv4 + ipv6 to simulate dual-stack
+	if ip.IsUnspecified() || splitLoopback {
+		ports = append(ports, port)
+		ports[0].Protocol = protocol + "4"
+		ports[1].Protocol = protocol + "6"
+		if splitLoopback {
+			// Hacky, but we will only have one ipv4 loopback with WSL config
+			ports[0].HostIP = "127.0.0.1"
```
Comment (Luap99) [inline libpod/networking_linux.go:114]: Should we really force `::1` to `127.0.0.1` IMO this is wrong when the users set a ip they want to only listen on that.

**Comment by n1hility** [comment]
@Luap99 hey, sorry I wasted your time looking at that last change. I didn't catch what should have been an obvious failure on cni/netavark. I rushed moving the code into port mapping and only did one basic rootful test, which was a false positive due to a stale networking config.

Thinking about this in more detail, I don't think it makes sense to apply this workaround across all network/port mapping implementations (explanation to follow). So I have switched my proposal back to the first/previous approach.

The WSL usability issue this change targets is purely around localhost network handling. They proxy directly from Windows localhost, for both IPv4 and IPv6, by discovering bound ports by address and then binding the same address+port pair on the Windows side. The defect in their approach is that they don't replicate dual-stack binds, and Windows defaults to off, so anything doing so will fail ipv4 connections. Unfortunately, since, as you know, the default on Linux (and everything else that chose to follow RF3493) is the opposite, anyone choosing to port map `::` will run into problems. 

Things really fall apart, though, when you combine this with Golang's networking behavior. While many impls would bind "0.0.0.0" (which we also convert to from an empty host IP) as AF_INET, Golang instead converts it to `::` using AF_INET6 with V6ONLY=false. With the behaviors combined, `0.0.0.0`, `::`, and an empty IP from a Golang-based service is unworkable on Windows. Note that since `0.0.0.0` and `::` translate to the same binding on Golang, attempting to bind both will trigger a socket error. To fix that, `0.0.0.0` must have a protocol of "tcp4", which will cause `net.Listen` to use `AF_INET`. 

With the workaround this PR introduced to rootlessport, both IPv4 and IPv6 function correctly,  mirroring the semantics of dual-stack binding. So, on WSL,`127.0.0.1` and `::1` work as expected.

Without any change, IPv4 works with cni/netavark rootfull networking since the port reservation code registers a dummy host binding and maps `0.0.0.0` to "tcp4". IPv6 localhost does not work, and can not work, in this mode since the kernel drops local scope packets before SNAT/MASQ can fire. Also, if you disable port reservation, nothing will work since WSL won't discover the port.

Also, without any change, IPv4 works with port_handler=slirp4netns since it is C-based and uses AF_INET. IPv6 is not supported at all with `add_hostfwd`; setting an IPv6 IP (e.g. `::`) will fail.

Long story short, the solution only makes sense and is needed for the rootless backend. For the others, it is either unnecessary or breaking. A possible complementary change would be to modify the cni/netavark port reservation to bind `0.0.0.0` and `127.0.0.1` whenever `::` or `::1` are set (respectively). This would allow a port mapping of `::` to work with IPv4 on localhost. However, this IMO is less important than getting IPv4 on rootless functioning.

File: libpod/networking_linux.go (hunk 1 above)
Comment (n1hility) [reply libpod/networking_linux.go]: We need that protocol distinction since otherwise you would end up with a port already bound error (since they value evaluate to the same binding on Golang). I moved the code back to rootlessport, so we know it will be handled correctly.  See the reply https://github.com/containers/podman/pull/14026#issuecomment-1113755530 for more details

The comma not being handled was an oversight in moving the code (in rootlessport it was already split). Now that its moved back this should also be resolved.

File: libpod/networking_linux.go (hunk 2 above)
Comment (n1hility) [reply libpod/networking_linux.go]: I mainly wanted to prevent a situation where if it was ever exposed since it would break the workaround. Although perhaps if it ever was supported it would be instead be implemented on a separate field. I am happy to delete this since its not part of the contract.

File: libpod/networking_linux.go (hunk 3 above)
Comment (n1hility) [reply libpod/networking_linux.go]: This code isn't forcing `::1` to `127.0.0.1`, it's adding 127.0.0.1 as the first binding, and ::1 as the second, which mirrors the behavior of a dual-stack `::1` bind. What I meant with the hacky comment was just that you can have multiple 127 addresses, and bindings would be missed on those.

**Review by Luap99 — APPROVED** [review]
LGTM

Thanks for your detailed description.
I really dislike that we have to fix broken wsl behaviour but here we are.

I agree now that this is the best way to fix it when iptables/and slirp4netns are not effected.

**Comment by n1hility** [comment]
> LGTM
> 
> Thanks for your detailed description. I really dislike that we have to fix broken wsl behaviour but here we are.
> 
> I agree now that this is the best way to fix it when iptables/and slirp4netns are not effected.

Thanks @Luap99. I hear you and promise to rip this code out the second WSL has addressed the problem.
